From 1f6046223a71057ab82ddfb47668b47082070546 Mon Sep 17 00:00:00 2001
From: Miroslav Benes <mbenes@suse.cz>
Date: Mon, 2 Jan 2017 15:43:01 +0100
Subject: [PATCH] kgraft/iscsi-target: Do not block kGraft in iscsi_np kthread
Patch-mainline: not yet, kgraft
References: bsc#1010612, fate#313296

The iscsi_np kthread may sleep interruptibly inside schedule_timeout()
called from inet_csk_wait_for_connect(). This is not really a place
where we could easily put a klp_kgraft_mark_task_safe() anotation. The
code path is even heavily shared with userspace, plus we're really deep
in a call chain.

Let's leave it on klp_kgraft_mark_task_safe(current) in
iscsi_target_login_thread() and break with -EAGAIN from the infinite
loop in inet_csk_wait_for_connect().

Reviewed-by: Petr Mladek <pmladek@suse.com>
Signed-off-by: Miroslav Benes <mbenes@suse.cz>
---
 net/ipv4/inet_connection_sock.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/net/ipv4/inet_connection_sock.c b/net/ipv4/inet_connection_sock.c
index 46b9c887bede..3fa3a5579a2f 100644
--- a/net/ipv4/inet_connection_sock.c
+++ b/net/ipv4/inet_connection_sock.c
@@ -292,6 +292,9 @@ static int inet_csk_wait_for_connect(struct sock *sk, long timeo)
 		err = -EAGAIN;
 		if (!timeo)
 			break;
+		if ((current->flags & PF_KTHREAD) &&
+		    klp_kgraft_task_in_progress(current))
+			break;
 	}
 	finish_wait(sk_sleep(sk), &wait);
 	return err;
-- 
2.11.0

